fix(hardware_control): Fixing liquid probe bugs found in end to end testing

ryanthecoder opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

These changes came from testing the end to end implementation of liquid level detection on the ot3 and fixing the bugs that came up.

# Test Plan

<!--
Use this section to describe the steps that you took to test your Pull Request.
If you did not perform any testing provide justification why.

OT-3 Developers: You should default to testing on actual physical hardware.
Once again, if you did not perform testing against hardware, justify why.

Note: It can be helpful to write a test plan before doing development

Example Test Plan (HTTP API Change)

- Verified that new optional argument `dance-party` causes the robot to flash its lights, move the pipettes,
then home.
- Verified that when you omit the `dance-party` option the robot homes normally
- Added protocol that uses `dance-party` argument to G-Code Testing Suite
- Ran protocol that did not use `dance-party` argument and everything was successful
- Added unit tests to validate that changes to pydantic model are correct

-->
Changed the unit tests to match new functionality.
Tested all new methods for liquid level detection on the ot3. 

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

Changes include:
* Deleting unnecessary checks in ot3controller.py
* Editing ot3api.py to issue the correct probe movements
* Standardizing the attribute names for liquid_presence_detection
* Conditionally adding a require_liquid_presence command before aspirates
* Redoing movement in liquid_probe so that it actually takes a sensible path
* Editing tests accordingly

# Review requests

<!--
Describe any requests for your reviewers here.
-->

We are aware of a bug that even though we handle LLD exceptions at the protocol_context level, protocol_engine commands will still be marked as failed, so all future commands will fail. As far as I know for this PR, this only affects detect_liquid_presence on a well with no liquid. Another PR will be coming soon that addresses this.

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

Medium? I'm not sure.

aaron-kulkarni committed:
testing fixes to lld

aaron-kulkarni committed:
fixed errors from yesterday. but well is being impaled

aaron-kulkarni committed:
minor things

ryanthecoder committed:
stuff

ryanthecoder committed:
fix the default config generator for data files

ryanthecoder committed:
remove our debug logging

sfoster1 commented on api/lldfunctionexamples.py:
```diff
@@ -0,0 +1,93 @@
+from opentrons import protocol_api, types
```
We shouldn't check this in, I think. I don't think this is a good location for it.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -1363,16 +1362,16 @@ async def liquid_probe(
-                raise UnsupportedHardwareCommand(
-                    "Liquid Probe not supported on this pipette firmware"
-                )
+        # if output_option == OutputOptions.sync_buffer_to_csv:
```
Delete rather than checking in commented otu code

sfoster1 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -851,6 +851,16 @@ def liquid_probe_with_recovery(self, well_core: WellCore) -> None:
             origin=WellOrigin.TOP, offset=WellOffset(x=0, y=0, z=0)
         )
 
+        loc = Location(well_core.get_top(10), "Well")
```
let's get this value from some config instead of it being a specific number literal

SyntaxColoring replied:
I'm also just noticing that our public `InstrumentContext` methods don't have a way of specifying an X/Y/Z offset, but the `liquidProbe` Protocol Engine command does (although [it's a work in progress](https://opentrons.slack.com/archives/C06M9F26YGH/p1719937084438019)).

Is the difference intentional?

CC @pmoegenburg since you were working on the Protocol Engine side.

SyntaxColoring replied:
`Location(well_core.get_top(10), "Well")` doesn't look like a valid way to construct a `Location`. It's trying to set the `.labware` attribute to a deck slot named `"Well"`. You might have intended to provide an actual instance of the `Well` class, or another value that's valid [per the types of this union](https://github.com/Opentrons/opentrons/blob/edge/api/src/opentrons/types.py#L111-L117).

Also keep in mind that `InstrumentContext` can be the thing that calls `InstrumentCore.move_to()`. I suspect that will be easier because you'll have better access to the things you'll need to build a `location` argument.

sfoster1 commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -867,6 +877,16 @@ def liquid_probe_without_recovery(self, well_core: WellCore) -> float:
             origin=WellOrigin.TOP, offset=WellOffset(x=0, y=0, z=0)
         )
 
+        loc = Location(well_core.get_top(10), "Well")
```
same here, but actually we can just reuse the `loc` value, it's the same well correct?

sfoster1 reviewed (dismissed):
Some style and consistency stuff to fix up. Also, let's get a nicer description in here.

SyntaxColoring commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -851,6 +851,16 @@ def liquid_probe_with_recovery(self, well_core: WellCore) -> None:
 
+        loc = Location(well_core.get_top(10), "Well")
+
+        self.move_to(
+            location=loc,
+            well_core=well_core,
+            force_direct=False,
+            minimum_z_height=None,
+            speed=None,
+        )
```
Why do we need to do this? Doesn't the `liquidProbe` Protocol Engine command already move to the target well?

SyntaxColoring replied:
We probably do need to update PAPI's logical "current well" bookkeeping, though. Like how `aspirate()` and `dispense()` do `self._protocol_core.set_last_location()`.

SyntaxColoring replied:
Discussed in-person. This is working around a weird problem with path planning, which we're looking into.

aaron-kulkarni replied:
Fixed, we were setting CurrentWell incorrectly in liquidProbe

aaron-kulkarni committed:
comment fix

aaron-kulkarni committed:
fixed movement error from tiprack to well

aaron-kulkarni committed:
mostly works except detect_liquid_presence errors on empty well instead of returning False

ryanthecoder commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -77,8 +77,9 @@ async def execute(self, params: LiquidProbeParams) -> _ExecuteReturn:
-            TipNotAttachedError: if there is not tip attached to the pipette
+            TipNotAttachedError: if there is no tip attached to the pipette
             MustHomeError: if the plunger is not in a valid position
+            TipNotEmptyError: if the tip starts with liquid in it(AKA it's wet and unusable for LLD)
```
So a tip can theoretically be wet and also empty, if we've used it before and then emptied it. So lets leave out that AKA part

aaron-kulkarni committed:
various fixes and improvements

aaron-kulkarni committed:
minor comment change

aaron-kulkarni committed:
minor lint stuff

aaron-kulkarni committed:
final? except for snapshot

aaron-kulkarni committed:
test fixes

aaron-kulkarni committed:
another test change

aaron-kulkarni committed:
formatting

aaron-kulkarni committed:
please let this be it

aaron-kulkarni committed:
ok this is actually it

aaron-kulkarni committed:
ok this is it

github-actions[bot] committed:
fix(analyses-snapshot-testing): fixing-liquid-probe snapshot failure capture (#15629)

This PR is an automated snapshot update request. Please review the
changes and merge if they are acceptable or find your bug and fix it.

Co-authored-by: aaron-kulkarni <107003644+aaron-kulkarni@users.noreply.github.com>

ryanthecoder commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -2055,12 +2062,14 @@ def detect_liquid_presence(self, well: labware.Well) -> bool:
         except ProtocolCommandFailedError as e:
-            if isinstance(e.original_error, LiquidNotFoundError):
+            # if we handle the error, we change the protocl state from error to valid
+            if isinstance(e.original_error, LiquidNotFoundError) or isinstance(
```
I'm not sure we should catch Stall's here, that's something that needs to be recovered from and if we've caused a stall then we've done something else wrong

SyntaxColoring commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -261,6 +260,14 @@ def aspirate(
+            and self.liquid_presence_detection
+        ):
+            self.require_liquid_presence(well=well)
+            self.prepare_to_aspirate()
```
Do we need this `self.prepare_to_aspirate()` call? It looks like this is [currently being handed in the `opentrons.protocol_engine` layer?](https://github.com/Opentrons/opentrons/blob/8b88dda01afa364464de3f1506763b1ec08adc6d/api/src/opentrons/protocol_engine/commands/aspirate.py#L104) Did you find that this was necessary for it to work?

aaron-kulkarni replied:
Yes, I thought it should be unnecessary too. But if I didn't have this prepare_to_aspirate, it was not aspirating any liquid at all after liquid probing.

SyntaxColoring replied:
OK, let's keep it for now to unblock testing, but let's keep looking into it. Might be worth leaving a `# todo` comment.

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -99,19 +100,12 @@ async def execute(self, params: LiquidProbeParams) -> _ExecuteReturn:
                 message="Current position of pipette is invalid. Please home."
             )
 
-        current_well = CurrentWell(
-            pipette_id=pipette_id,
-            labware_id=labware_id,
-            well_name=well_name,
-        )
-
```
Ah, yep, that'd do it. Nice catch.

SyntaxColoring commented on api/src/opentrons/protocol_engine/execution/pipetting.py:
```diff
@@ -82,7 +82,7 @@ def __init__(self, state_view: StateView, hardware_api: HardwareControlAPI) -> N
 
     def get_is_empty(self, pipette_id: str) -> bool:
         """Get whether a pipette has a working volume equal to 0."""
-        return self._state_view.pipettes.get_working_volume(pipette_id) == 0
+        return self._state_view.pipettes.get_aspirated_volume(pipette_id) == 0
```
Does the docstring here need to be updated?

aaron-kulkarni replied:
Yes, thanks for noticing that.

SyntaxColoring commented on api/src/opentrons/protocol_engine/execution/pipetting.py:
```diff
@@ -235,7 +235,7 @@ def __init__(
 
     def get_is_empty(self, pipette_id: str) -> bool:
         """Get whether a pipette has a working volume equal to 0."""
-        return self._state_view.pipettes.get_working_volume(pipette_id) == 0
+        return self._state_view.pipettes.get_aspirated_volume(pipette_id) == 0
```
Does the docstring here need to be updated?

aaron-kulkarni committed:
don't catch stalls in detect_liquid_presence

SyntaxColoring reviewed (approved):
This all looks good to me assuming others' pending comments have been resolved. Some small final notes:

aaron-kulkarni committed:
addressing comments
